fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/Quantifier.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * Quantifier.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2020 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.temp;
+
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryPlan;
+import com.apple.foundationdb.record.query.plan.temp.matchers.ExpressionMatcher;
+import com.apple.foundationdb.record.query.plan.temp.matchers.PlannerBindings;
+import com.google.common.collect.ImmutableList;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+/**
+ * A quantifier is a conduit facilitating the data flow between the output of one {@link RelationalExpression} R and the
+ * consumption of that data by another {@link RelationalExpression} S. S is said to own the quantifier, while the
+ * quantifier is said to range over R. Quantifiers come in very few but very distinct flavors. All flavors are
+ * implemented by static inner final classes as to emulate a sealed trait.
+ *
+ * Quantifiers separate what it means to be producing versus consuming records. The expression a quantifier ranges over
+ * produces records, the quantifier flows information (according to the flavor) which is then consumed by the expression
+ * containing or owning the quantifier. That expression can consume the data in a way independent of how the data was
+ * produced in the first place.
+ *
+ * A quantifier works closely with the expression that owns it. Depending on the semantics of the owning expression
+ * it becomes possible to model correlations,  e.g. for a logical join expression the quantifier can provide a binding
+ * of the record being currently consumed by the join's outer to other (inner) parts of the data flow that are also rooted
+ * at the owning (join) expression.
+ */
+public abstract class Quantifier implements Bindable {
+    /**
+     * For Each quantifier. Conceptually flows one record at a time from the expression it ranges over to
+     * the owning expression.
```

Comment:
I found the phrasing on these javadocs confusing, specifically the beginning part that is not really a sentence really threw me off and seems unnecessary. 
```suggestion
     * A quantifier that conceptually flows one record at a time from the expression it ranges over to
     * the owning expression.
```